From 2be2212c9a813425fd4be4ecc1ba5693f46ee947 Mon Sep 17 00:00:00 2001
From: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Date: Wed, 7 Nov 2012 15:50:27 -0800
Subject: Dalvik: Disassembler functions do not show all the arguments

BZ: 66992

Disassembler functions do not show all the arguments

Fixes the hidden arguments. The len + left parameters were not always updated as they should have been. The patch fixes that.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I4934bd5ab496e30093576e433f9025d913dd01a8
Orig-MCG-Change-Id: Iae5dffd68e8d83aba89ee5f35012a177c512473f
Signed-off-by: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/Dataflow.cpp |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/Dataflow.cpp b/vm/compiler/Dataflow.cpp
index fa8ae8a..0707bdc 100644
--- a/vm/compiler/Dataflow.cpp
+++ b/vm/compiler/Dataflow.cpp
@@ -892,13 +892,21 @@ char *dvmCompilerGetDalvikDisassembly(const DecodedInstruction *insn,
     else {
         if (dfAttributes & DF_A_IS_REG) {
             snprintf(buffer + len, left, " v%d", insn->vA);
+            len = strlen (buffer);
+            left = sizeof (buffer) - len;
         }
+
         if (dfAttributes & DF_B_IS_REG) {
             snprintf(buffer + len, left, ", v%d", insn->vB);
         }
         else if ((int)opcode < (int)kMirOpFirst) {
             snprintf(buffer + len, left, ", (#%d)", insn->vB);
         }
+
+        //Recalculate
+        len = strlen (buffer);
+        left = sizeof (buffer) - len;
+
         if (dfAttributes & DF_C_IS_REG) {
             snprintf(buffer + len, left, ", v%d", insn->vC);
         }
@@ -1136,6 +1144,8 @@ char *dvmCompilerFullDisassembler(const CompilationUnit *cUnit,
                 dvmAbort();
                 break;
         }
+        offset = strlen (buffer);
+        left = sizeof (buffer) - offset;
         snprintf(buffer + offset, left, " %04x",
                  mir->offset + delta);
     } else if (dfAttributes & (DF_FORMAT_35C | DF_FORMAT_3RC)) {
-- 
1.7.4.1

